File: Week8/cucumber-exercises/src/test/java/com/training/cucumber/stepdefinitions/LoginSteps.java
package com.training.cucumber.stepdefinitions;

import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;

import static org.junit.jupiter.api.Assertions.*;

public class LoginSteps {

    private WebDriver driver;
    private static final String BASE_URL = "https://the-internet.herokuapp.com";

    @After
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }

    @Given("the user is on the login page")
    public void theUserIsOnTheLoginPage() {
        // 1. Set up WebDriverManager for Chrome
        // 2. Initialize ChromeDriver
        // 3. Navigate to login page

        WebDriverManager.chromedriver().setup();
        driver = new ChromeDriver();
        driver.get(BASE_URL + "/login");
    }

    @When("the user enters username {string}")
    public void theUserEntersUsername(String username) {
        // Find username field and enter the provided username

        WebElement usernameField = driver.findElement(By.id("username"));
        usernameField.sendKeys(username);
    }

    @When("the user enters password {string}")
    public void theUserEntersPassword(String password) {
        // Find password field and enter the provided password

        driver.findElement(By.id("password")).sendKeys(password);
    }

    @When("the user clicks the login button")
    public void theUserClicksTheLoginButton() {
        // Find and click the login button

        driver.findElement(By.cssSelector(".fa.fa-2x.fa-sign-in")).click();
    }

    @Then("the user should be redirected to the secure area")
    public void theUserShouldBeRedirectedToTheSecureArea() {
        // Verify the URL contains "/secure"

        assertTrue(driver.getCurrentUrl().contains("/secure"),
                "User was not redirected to secure area");
    }

    @Then("the user should see a success message containing {string}")
    public void theUserShouldSeeSuccessMessageContaining(String expectedMessage) {
        // Find the flash message element and verify it contains expected text

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        WebElement flash = wait.until(
                ExpectedConditions.visibilityOfElementLocated(By.id("flash"))
        );
        String actualMessage = flash.getText();
        assertTrue(actualMessage.contains(expectedMessage),
                "User did not see a success message containing " + expectedMessage);
    }

    @Then("the user should remain on the login page")
    public void theUserShouldRemainOnTheLoginPage() {
        // Verify URL still contains "/login"

        assertTrue(driver.getCurrentUrl().contains("/login"),
                "User did not remain on the login page");
    }

    @Then("the user should see an error message containing {string}")
    public void theUserShouldSeeErrorMessageContaining(String expectedMessage) {
        // Find the flash message element and verify it contains expected error

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        WebElement flash = wait.until(
                ExpectedConditions.visibilityOfElementLocated(By.id("flash"))
        );
        String actualMessage = flash.getText();
        assertTrue(actualMessage.contains(expectedMessage),
                "User did not see an error message containing " + expectedMessage);
    }
}

File: Week8/PythonBehave/project/features/steps/login_steps.py
from behave import given, when, then
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as ec
from selenium.webdriver.support.wait import WebDriverWait

BASE_URL = 'https://the-internet.herokuapp.com/'

@given(u'the application is running')
def application_running(context):
    options = Options()
    options.add_argument('--start-maximized')

    context.driver = webdriver.Chrome(options=options)
    context.wait = WebDriverWait(context.driver, 10)

    context.driver.get(BASE_URL)

    heading = context.wait.until(ec.visibility_of_element_located((By.TAG_NAME, 'h1')))
    assert heading.text == 'Welcome to the-internet'

@given(u'the test database is already seeded with users')
def database_seeded_with_users(context):
    return True

@given(u'the user is on the login page')
def user_on_login_page(context):
    context.driver.get(BASE_URL + '/login')
    heading = context.wait.until(ec.visibility_of_element_located((By.TAG_NAME, 'h2')))
    assert 'Login Page' in heading.text

@when(u'the user enters username "{username}"')
def user_enters_username(context, username):
    username_input = context.wait.until(ec.visibility_of_element_located((By.ID, 'username')))
    username_input.clear()
    username_input.send_keys(username)

@when(u'the user enters password "{password}"')
def user_enters_password(context, password):
    password_input = context.wait.until(ec.visibility_of_element_located((By.ID, 'password')))
    password_input.clear()
    password_input.send_keys(password)

@when(u'the user clicks the login button')
def user_clicks_login(context):
    login_button = context.wait.until(ec.visibility_of_element_located((By.CSS_SELECTOR, '.fa.fa-2x.fa-sign-in')))
    login_button.click()

@then(u'the user should be redirected to the secure area')
def step_impl(context):
    context.wait.until(ec.url_contains('/secure'))
    assert '/secure' in context.driver.current_url

@then(u'the page should display a message containing "{message_text}"')
def step_impl(context, message_text):
    flash_element = context.wait.until(ec.visibility_of_element_located((By.ID, 'flash')))
    assert message_text in flash_element.text